include/deal.II/lac/constraint_linear_operator.h

```diff
@@ -0,0 +1,291 @@
+  {
+    v = 0.;
+    Tvmult_add(v, u);
+  };
+
+  return return_op;
+}
+
+
+/**
+ * Given a ConstraintMatrix @p constraint_matrix and an operator exemplar
+ * @p exemplar, return an LinearOperator that is the projection to the
+ * subspace of constrained degrees of freedom.
+ *
+ * @author Mauro Bardelloni, Matthias Maier, 2015
+ *
+ * @relates LinearOperator
+ * @ingroup constraints
+ */
+template <typename Range, typename Domain>
+LinearOperator<Range, Domain> project_to_constrained_linear_operator(
+  const ConstraintMatrix &constraint_matrix,
+  const LinearOperator<Range, Domain> &exemplar)
+{
+  LinearOperator<Range, Domain> return_op = exemplar;
+
+  return_op.vmult_add = [&constraint_matrix](Range &v, const Domain &u)
+  {
+    for (auto i : v.locally_owned_elements())
+      if (constraint_matrix.is_constrained(i))
+        v(i) += u(i);
+  };
+
+  return_op.Tvmult_add = [&constraint_matrix](Domain &v, const Range &u)
+  {
+    for (auto i : v.locally_owned_elements())
+      if (constraint_matrix.is_constrained(i))
+        v(i) += u(i);
+  };
+
+  return_op.vmult = [&constraint_matrix](Range &v, const Domain &u)
+  {
+    v = 0.;
+    for (auto i : v.locally_owned_elements())
+      if (constraint_matrix.is_constrained(i))
+        v(i) += u(i);
+  };
+
+  return_op.Tvmult = [&constraint_matrix](Domain &v, const Range &u)
+  {
+    v = 0.;
+    for (auto i : v.locally_owned_elements())
+      if (constraint_matrix.is_constrained(i))
+        v(i) += u(i);
+  };
+
+  return return_op;
+}
+
+
+/**
+ * Given a ConstraintMatrix object @p constraint_matrix and a
+ * LinearOperator @p linop, this function creates a LinearOperator object
+ * consisting of the composition of three operations and a regularization:
+ * @code
+ *   Ct * linop * C + Id_c;
+ * @endcode
+ * with
+ * @code
+ *   C = distribute_constraints_linear_operator(constraint_matrix, linop);
+ *   Ct = transpose_operator(C);
+ *   Id_c = project_to_constrained_linear_operator(constraint_matrix, linop);
+ * @endcode
+ * and <code>Id_c</code> is the projection to the subspace consisting of
+ * all vector entries associated with constrained degrees of freedoms.
+ *
+ * This LinearOperator object is used together with
+ * constrained_right_hand_side() to build up the following modified system
+ * of linear equations:
+ * @f[
+ *   (C^T A C + Id_c) x = C^T (b - A\,k)
+ * @f]
+ * with a given (unconstrained) system matrix $A$, right hand side $b$, and
+ * linear constraints $C$ with inhomogeneities $k$.
+ *
+ * A detailed explanation of this approach is given in the @ref constraints
+ * module.
+ *
+ * @author Mauro Bardelloni, Matthias Maier, 2015
+ *
+ * @relates LinearOperator
+ * @ingroup constraints
+ */
+template <typename Range, typename Domain>
+LinearOperator<Range, Domain>
+constrained_linear_operator(const ConstraintMatrix &constraint_matrix,
+                            const LinearOperator<Range, Domain> &linop)
```

Comment:
the function is called `constrained_linear_operator`, but the filename is `constraint_linear_operator.h`. Is this on purpose?